src/server/renderToStream/orchestrateChunks.ts

```diff
@@ -0,0 +1,120 @@
+  let lastWritePromise: null | Promise<void> = null
+
+  // See Rule 2: https://github.com/brillout/react-streaming/tree/main/src#rule-2
+  let firstReactWritePromise_resolve: () => void
+  let firstReactWritePromise: Promise<void> | null = new Promise<void>((resolve) => {
+    firstReactWritePromise_resolve = () => {
+      if (firstReactWritePromise === null) return
+      firstReactWritePromise = null
+      resolve()
+    }
+  })
+  let isFirstReactWrite = true
+
+  return { injectToStream, onReactWrite, onBeforeEnd, hasStreamEnded: () => hasEnded }
+
+  function injectToStream(chunk: Chunk, options?: InjectToStreamOptions) {
+    if (debug.isEnabled) debug('injectToStream()', getChunkAsString(chunk))
+    if (hasEnded) {
+      assertUsage(
+        false,
+        `Cannot inject the following chunk because the stream has already ended. Consider using the doNotClose() and hasStreamEnded() utilities. The chunk:\n${getChunkAsString(
+          chunk,
+        )}`,
+      )
+    }
+    writeChunkInSequence(chunk, options?.flush)
+  }
+
+  // Except of the first React chunk, all chunks are guaranteed to be written in the
+  // order of the injectToStream() and onReactWrite() calls.
+  function writeChunkInSequence(chunk: unknown, flush?: boolean) {
+    const lastWritePromisePrevious = lastWritePromise
+    lastWritePromise = (async () => {
+      if (firstReactWritePromise) await firstReactWritePromise
+      if (lastWritePromisePrevious) await lastWritePromisePrevious
+      if (isPromise(chunk)) chunk = await chunk
+      writeChunkNow(chunk, flush)
+    })()
+  }
+  function writeChunkNow(chunk: unknown, flush?: boolean) {
+    assert(!hasEnded)
+    assert(streamOperations.operations)
+
+    // Write
+    streamOperations.operations.writeChunk(chunk)
+    if (debug.isEnabled) debug('>>> WRITE', getChunkAsString(chunk))
+
+    // Flush
+    if (flush && streamOperations.operations.flush !== null) {
+      streamOperations.operations.flush()
+      debug('>>> FLUSH')
+    }
+  }
+
+  function onReactWrite(chunk: unknown) {
+    if (debug.isEnabled) debug('onReactWrite()', getChunkAsString(chunk))
+    assert(!hasEnded)
+    const flush = true
+    if (isFirstReactWrite) {
+      isFirstReactWrite = false
+      debug('>>> START')
+      // The first React chunk is always the very first written chunk, see Rule 2:
+      // https://github.com/brillout/react-streaming/tree/main/src#rule-2
+      writeChunkNow(chunk, flush)
+      firstReactWritePromise_resolve()
+    } else {
+      writeChunkInSequence(chunk, flush)
+    }
+  }
+
+  async function onBeforeEnd() {
+    // In case React didn't write anything
+    firstReactWritePromise_resolve()
+    // Ensure user is able to use doNotClose() because, otherwise, stream may already have ended after `const { doNotClose } = await renderToStream()`
+    await new Promise<void>((r) => setTimeout(r, 0))
+    await doNotClosePromise.promise
+    hasEnded = true
+    await lastWritePromise
```

Comment:
```
    hasEnded = true
    await lastWritePromise
 ```
Shouldn't these be swapped? The `assert(!hasEnded)` can fail in `await lastWritePromise`

Reply:
9a4c4e153ba6f4e1bdd218f2987339deecbc254e